flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -318,6 +308,20 @@ private void isReadyForDeployment(ClusterSpecification clusterSpecification) thr
 		}
 	}
 
+	@VisibleForTesting
+	protected int getNumYarnMaxVcores() throws YarnDeploymentException {
```

Comment:
I am wondering whether it is actually less invasive to introduce a `StubYarnClientImpl` and use in `YarnClusterDescriptorTest#setup` instead of changing production class `YarnClusterDescriptor` for this:
```
class StubYarnClientImpl() extends YarnClient {
			@Override
			public List<NodeReport> getNodeReports(NodeState... states) {
				return Collections.singletonList(new NodeReport() {
					@Override
					public Resource getCapability() {
						return new Resource() {
							@Override
							public int getVirtualCores() {
								return NUM_YARN_MAX_VCORES;
							}
							// ...
						};
					}
					// ...
				});
			}
            // ...
}

public class YarnClusterDescriptorTest extends TestLogger {
    // ..
    @BeforeClass
	public static void setupClass() {
		yarnConfiguration = new YarnConfiguration();
		yarnClient = new StubYarnClientImpl();
		yarnClient.init(yarnConfiguration);
		yarnClient.start();
	}
    //..
}
```

It is quite some useless code but we can put into a separate file.

Reply:
I'm not sure about this.

`YarnClient` is an abstract class, and to introduce a `StubYarnClientImpl` we would need to also implement more than 20 useless methods and mock the `NodeReport` as well. The complication seems unnecessary to me.

It is true the current approach touches the production class `YarnClusterDescriptor`, but only with a trivial refactor. IMO, even without this testability issue, it is not a bad thing to extract the logic getting max vcores from Yarn into a separate method.

Reply:
I agree with Andrey that overriding production code methods should always be our last resort when it comes to testing. The danger is that this method evolves and that we override important behaviour unintentionally. I would propose two solutions to the problem:

1. Introduce a `YarnClusterInformationRetriever` interface which offers the method `getMaxVcores`. The default implementation will simply use the `YarnClient` to retrieve the max vcores. In the test we can provide a testing implementation.
2. Alternatively, similar to `TestingYarnClient` override the `getNodeReports` method from `YarnClientImpl`.

Reply:
I also looked into `YarnClientImpl`. My only concern about this option was that it is annotated  with `@Private` and `@Unstable` but I did not see that we actually already decided once to extend it. I would be ok with it. I guess we will refactor it if this class is gone after updating the Yarn dependency.